test/functional/feature_utxo_set_hash.py

```diff
@@ -29,39 +34,29 @@ def test_muhash_implementation(self):
         mocktime = node.getblockheader(node.getblockhash(0))['time'] + 1
         node.setmocktime(mocktime)
 
-        # Generate 100 blocks and remove the first since we plan to spend its
-        # coinbase
-        block_hashes = self.generate(wallet, 1) + self.generate(node, 99)
-        blocks = list(map(lambda block: from_hex(CBlock(), node.getblock(block, False)), block_hashes))
-        blocks.pop(0)
-
-        # Create a spending transaction and mine a block which includes it
+        # Generate 100 blocks, create spending transaction and mine another block
+        self.generate(wallet, 1) + self.generate(node, 99)
         txid = wallet.send_self_transfer(from_node=node)['txid']
-        tx_block = self.generateblock(node, output=wallet.get_address(), transactions=[txid])
-        blocks.append(from_hex(CBlock(), node.getblock(tx_block['hash'], False)))
+        self.generateblock(node, output=wallet.get_address(), transactions=[txid])
 
-        # Serialize the outputs that should be in the UTXO set and add them to
-        # a MuHash object
+        # Serialize created/spent transaction outputs in each block and
+        # apply (add/remove) them to the MuHash object accordingly
         muhash = MuHash3072()
 
-        for height, block in enumerate(blocks):
-            # The Genesis block coinbase is not part of the UTXO set and we
-            # spent the first mined block
-            height += 2
-
-            for tx in block.vtx:
-                for n, tx_out in enumerate(tx.vout):
-                    coinbase = 1 if not tx.vin[0].prevout.hash else 0
-
-                    # Skip witness commitment
-                    if (coinbase and n > 0):
-                        continue
-
-                    data = COutPoint(int(tx.rehash(), 16), n).serialize()
-                    data += struct.pack("<i", height * 2 + coinbase)
-                    data += tx_out.serialize()
-
-                    muhash.insert(data)
+        for height in range(1, node.getblockcount() + 1):
+            for tx in node.getblock(blockhash=node.getblockhash(height), verbose=3)['tx']:
+                # add created coins
+                is_coinbase_tx = 'coinbase' in tx['vin'][0]
+                for o in tx['vout']:
+                    if o['scriptPubKey']['type'] != 'nulldata':
+                        muhash.insert(txout_serialize(tx['txid'], o['n'], height, is_coinbase_tx,
+                                                      o['value'], o['scriptPubKey']['hex']))
+                # remove spent coins
+                if not is_coinbase_tx:
+                    for i in tx['vin']:
```

Comment:
This doesn't seem appropriate, since `vin` (and `vout`) doesn't denote a single input/output, but rather a collection of them. I assume the `v` prefix stems from `vector`, see e.g. https://github.com/bitcoin/bitcoin/blob/9c47eb450346937b3d7ee21b9e669b5c91a7ed6c/src/primitives/transaction.h#L298-L299
I was thinking only using `in`/`out`, but `in` is a reserved word in Python (`for in in ...` would look weird), `input`/`output` is IMHO too long, so I decided to just stick with `i`/`o`.

Reply:
You're absolutely right about `vin`/`vout` being inappropriate for a single input/output, my bad. I don't agree with `input`/`output` being too long though, especially here when we have multiple layers of nesting. However, it's also not important enough and probably too much personal preference to bikeshed over so I'll leave it up to you.